edge/pkg/devicetwin/dtcommon/util.go

```diff
@@ -32,7 +32,7 @@ func ValidateValue(valueType string, value string) error {
 			return errors.New("the bool value must be true or false")
 		}
 		return nil
-	case "deleted":
+	case TypeDeleted:
```

Comment:
Why only add the `TypeDeleted` here? :) And the original code is also confusing, the operation type mixed with value type...

Reply:
yes, I check the code, this `MetaType deleted`, used for`cloud` notification `edge` delete the twin attributes.

 It can be found in the `downstream.go` line 660-690 in the cloud deviceController
```go
// addDeletedTwins add deleted twins in the message
func addDeletedTwins(oldTwin []v1alpha2.Twin, newTwin []v1alpha2.Twin, twin map[string]*types.MsgTwin, version string) {
	opt := false
	optional := &opt
	for i, dtwin := range oldTwin {
		if !ifTwinPresent(dtwin, newTwin) {
			expected := &types.TwinValue{}
			expected.Value = &oldTwin[i].Desired.Value
			timestamp := time.Now().UnixNano() / 1e6

			metadata := &types.ValueMetadata{Timestamp: timestamp}
			expected.Metadata = metadata

			// TODO: how to manage versioning ??
			cloudVersion, err := strconv.ParseInt(version, 10, 64)
			if err != nil {
				klog.Warningf("Failed to parse cloud version due to error %v", err)
			}
			twinVersion := &types.TwinVersion{CloudVersion: cloudVersion, EdgeVersion: 0}
			msgTwin := &types.MsgTwin{
				Expected:        expected,
				Optional:        optional,
				Metadata:        &types.TypeMetadata{Type: "deleted"},
				ExpectedVersion: twinVersion,
			}
			twin[dtwin.PropertyName] = msgTwin
		}
	}
}
```

I think it may be necessary to reorganize the `device communication` between `cloud` and `edge` in the future.